packages/contracts-bedrock/test/L2/L2CrossDomainMessenger.t.sol

```diff
@@ -237,4 +228,136 @@ contract L2CrossDomainMessenger_Test is Bridge_Initializer {
+        assertTrue(l2CrossDomainMessenger.isCustomGasToken());
+
+        bytes memory xDomainCallData =
+            Encoding.encodeCrossDomainMessage(l2CrossDomainMessenger.messageNonce(), alice, recipient, 0, 100, hex"ff");
+        vm.expectCall(
+            address(l2ToL1MessagePasser),
+            abi.encodeWithSelector(
+                L2ToL1MessagePasser.initiateWithdrawal.selector,
+                address(l1CrossDomainMessenger),
+                l2CrossDomainMessenger.baseGas(hex"ff", 100),
+                xDomainCallData
+            )
+        );
+
+        // MessagePassed event
+        vm.expectEmit(true, true, true, true);
+        emit MessagePassed(
+            l2ToL1MessagePasser.messageNonce(),
+            address(l2CrossDomainMessenger),
+            address(l1CrossDomainMessenger),
+            0,
+            l2CrossDomainMessenger.baseGas(hex"ff", 100),
+            xDomainCallData,
+            Hashing.hashWithdrawal(
+                Types.WithdrawalTransaction({
+                    nonce: l2ToL1MessagePasser.messageNonce(),
+                    sender: address(l2CrossDomainMessenger),
+                    target: address(l1CrossDomainMessenger),
+                    value: 0,
+                    gasLimit: l2CrossDomainMessenger.baseGas(hex"ff", 100),
+                    data: xDomainCallData
+                })
+            )
+        );
+
+        vm.prank(alice);
+        l2CrossDomainMessenger.sendMessage(recipient, hex"ff", uint32(100));
+    }
+
+    /// @dev Tests that the sendMessage reverts when call value is non-zero with custom gas token.
+    function test_sendMessage_customGasToken_withValue_reverts() external {
+        // Mock the gasPayingToken function to return a custom gas token
+        vm.mockCall(address(l1Block), abi.encodeWithSignature("gasPayingToken()"), abi.encode(address(1), uint8(2)));
+        assertTrue(l2CrossDomainMessenger.isCustomGasToken());
+
+        vm.expectRevert("CrossDomainMessenger: cannot send value with custom gas token");
+        l2CrossDomainMessenger.sendMessage{ value: 1 }(recipient, hex"ff", uint32(100));
+    }
+
+    /// @dev Tests that the relayMessage succeeds with a custom gas token when the call value is zero.
+    function test_relayMessage_customGasToken_noValue_succeeds() external {
+        // Mock the gasPayingToken function to return a custom gas token
+        vm.mockCall(address(l1Block), abi.encodeWithSignature("gasPayingToken()"), abi.encode(address(1), uint8(2)));
+        assertTrue(l2CrossDomainMessenger.isCustomGasToken());
+
+        address target = address(0xabcd);
+        address sender = address(l1CrossDomainMessenger);
+        address caller = AddressAliasHelper.applyL1ToL2Alias(address(l1CrossDomainMessenger));
+
+        vm.expectCall(target, hex"1111");
+
+        vm.prank(caller);
+
+        vm.expectEmit(true, true, true, true);
```

Comment:
nit: you can use `vm.expectEmit(address)` over the one with 4 bools, it will assume bools as true and also enforce that the event comes from the address that is passed in as an arg